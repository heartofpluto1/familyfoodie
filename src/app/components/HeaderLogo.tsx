'use client';

import Link from 'next/link';
import Image from 'next/image';
import { signOut } from 'next-auth/react';
import { LogoutIcon, BurgerIcon } from './Icons';
import { useRef, useEffect, useState } from 'react';
import UserSettings from './UserSettings';
import type { Session } from 'next-auth';

interface HeaderLogoProps {
	session: Session | null;
}

const HeaderLogo = ({ session }: HeaderLogoProps) => {
	const detailsRef = useRef<HTMLDetailsElement>(null);
	const [isUserSettingsOpen, setIsUserSettingsOpen] = useState(false);
	const isAuthenticated = !!session;

	const closeMenu = () => {
		if (detailsRef.current) {
			detailsRef.current.open = false;
		}
	};

	useEffect(() => {
		const handleAnyClick = (event: MouseEvent) => {
			if (detailsRef.current && detailsRef.current.open) {
				// Close menu on any click, except on the summary button itself
				if (!detailsRef.current.querySelector('summary')?.contains(event.target as Node)) {
					closeMenu();
				}
			}
		};

		document.addEventListener('click', handleAnyClick);
		return () => {
			document.removeEventListener('click', handleAnyClick);
		};
	}, []);

	return (
		<header className="bg-surface border-b border-custom">
			<div className="container mx-auto px-4 py-3 md:py-4">
				<div className="flex items-center justify-between">
					{/* Title Section */}
					<div className="min-w-0 flex-shrink">
						<h1 className="text-xl sm:text-2xl md:text-3xl text-foreground tracking-wide flex items-center gap-2">
							Family Foodie
							<span className="text-xs bg-blue-600 text-white px-2 py-0.5 rounded-sm font-medium">BETA</span>
						</h1>
						<p className="text-sm text-muted font-light italic">What the fork is for dinner?</p>
					</div>

					{/* Navigation and Auth */}
					<div className="flex items-center space-x-2 sm:space-x-4 md:space-x-6">
						{/* Navigation - only show when authenticated */}
						{isAuthenticated && (
							<nav className="hidden sm:block">
								<div className="flex space-x-3 md:space-x-6">
									<Link
										href="/"
										className="text-secondary hover:text-foreground transition-colors font-medium underline-offset-4 hover:underline text-sm md:text-base"
									>
										Home
									</Link>
									<Link
										href="/plan"
										className="text-secondary hover:text-foreground transition-colors font-medium underline-offset-4 hover:underline text-sm md:text-base"
									>
										Plan
									</Link>
									<Link
										href="/shop"
										className="text-secondary hover:text-foreground transition-colors font-medium underline-offset-4 hover:underline text-sm md:text-base"
									>
										Shop
									</Link>
									<Link
										href="/recipes"
										className="text-secondary hover:text-foreground transition-colors font-medium underline-offset-4 hover:underline text-sm md:text-base"
									>
										Recipes
									</Link>
									{session?.user?.is_admin && (
										<Link
											href="/admin"
											className="transition-colors font-medium underline-offset-4 hover:underline text-sm md:text-base hidden lg:inline"
										>
											Admin
										</Link>
									)}
								</div>
							</nav>
						)}

						{/* Mobile Navigation Menu - show only on small screens when authenticated */}
						{isAuthenticated && (
							<nav className="sm:hidden">
								<details ref={detailsRef} className="relative">
									<summary className="list-none cursor-pointer bg-surface border border-custom rounded-sm p-2 text-foreground hover:bg-accent/10 transition-colors flex items-center justify-center [&::-webkit-details-marker]:hidden">
										<BurgerIcon className="w-4 h-4" />
									</summary>
									<div className="absolute right-0 top-full mt-1 bg-surface border border-custom rounded-sm shadow-lg min-w-24 z-50">
										<Link href="/" className="block px-3 py-2 text-sm text-secondary hover:text-foreground hover:bg-accent/10 transition-colors">
											Home
										</Link>
										<Link href="/plan" className="block px-3 py-2 text-sm text-secondary hover:text-foreground hover:bg-accent/10 transition-colors">
											Plan
										</Link>
										<Link href="/shop" className="block px-3 py-2 text-sm text-secondary hover:text-foreground hover:bg-accent/10 transition-colors">
											Shop
										</Link>
										<Link
											href="/recipes"
											className="block px-3 py-2 text-sm text-secondary hover:text-foreground hover:bg-accent/10 transition-colors"
										>
											Recipes
										</Link>
										{session?.user?.is_admin && (
											<Link href="/admin" className="block px-3 py-2 text-sm transition-colors">
												Admin
											</Link>
										)}
									</div>
								</details>
							</nav>
						)}

						{/* Auth Section */}
						<div className={isAuthenticated ? 'border-l border-custom pl-2 sm:pl-4' : ''}>
							{isAuthenticated ? (
								<div className="flex items-center space-x-2 sm:space-x-3">
									<button
										onClick={() => setIsUserSettingsOpen(true)}
										className="flex items-center justify-center cursor-pointer"
										title="User settings"
									>
										{session?.user?.image ? (
											<Image
												src={session.user.image}
												alt={session.user.name || 'User'}
												width={36}
												height={36}
												className="w-8 h-8 sm:w-9 sm:h-9 rounded-full border border-custom"
												unoptimized
											/>
										) : (
											<div className="w-8 h-8 sm:w-9 sm:h-9 rounded-full bg-accent text-background flex items-center justify-center text-sm font-medium">
												{session?.user?.name?.charAt(0)?.toUpperCase() || 'U'}
											</div>
										)}
									</button>
									<button onClick={() => signOut({ callbackUrl: '/' })} className="btn-default p-1.5 sm:p-2 rounded-sm inline-block" title="Sign Out">
										<LogoutIcon className="w-4 h-4 sm:w-5 sm:h-5" />
									</button>
								</div>
							) : (
								<Link
									href="/auth/signin"
									className="bg-accent text-background px-3 py-1.5 sm:px-4 sm:py-2 rounded-sm text-xs sm:text-sm font-medium hover:bg-accent/90 transition-colors"
								>
									Sign In
								</Link>
							)}
						</div>
					</div>
				</div>
			</div>

			{/* User Settings Panel */}
			<UserSettings isOpen={isUserSettingsOpen} onClose={() => setIsUserSettingsOpen(false)} />
		</header>
	);
};

export default HeaderLogo;
