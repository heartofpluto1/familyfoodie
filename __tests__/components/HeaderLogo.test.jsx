import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import HeaderLogo from '@/app/components/HeaderLogo' // Adjust path as needed

// Mock Next.js Link component
jest.mock('next/link', () => {
  return function MockLink({ children, href, className }) {
    return (
      <a href={href} className={className} data-testid={`link-${href}`}>
        {children}
      </a>
    )
  }
})

// Mock Next.js font
jest.mock('next/font/google', () => ({
  Crimson_Text: jest.fn(() => ({
    className: 'mocked-crimson-text-font',
  })),
}))

describe('HeaderLogo Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('Basic Rendering', () => {
    it('renders without crashing', () => {
      render(<HeaderLogo />)
    })

    it('renders as semantic header element', () => {
      render(<HeaderLogo />)
      
      const headerElement = screen.getByRole('banner')
      expect(headerElement).toBeInTheDocument()
      expect(headerElement.tagName).toBe('HEADER')
    })

    it('has correct CSS classes on header element', () => {
      render(<HeaderLogo />)
      
      const headerElement = screen.getByRole('banner')
      expect(headerElement).toHaveClass(
        'bg-surface',
        'border-b',
        'border-custom'
      )
    })
  })

  describe('Title Section', () => {
    it('renders the main title correctly', () => {
      render(<HeaderLogo />)
      
      const title = screen.getByRole('heading', { level: 1 })
      expect(title).toHaveTextContent('Family Foodie')
      expect(title).toBeInTheDocument()
    })

    it('applies correct font classes to title', () => {
      render(<HeaderLogo />)
      
      const title = screen.getByRole('heading', { level: 1 })
      expect(title).toHaveClass(
        'mocked-crimson-text-font',
        'font-bold',
        'text-4xl',
        'md:text-5xl',
        'text-foreground',
        'mb-2',
        'tracking-wide'
      )
    })

    it('renders the subtitle correctly', () => {
      render(<HeaderLogo />)
      
      const subtitle = screen.getByText('What the fork is for dinner?')
      expect(subtitle).toBeInTheDocument()
      expect(subtitle.tagName).toBe('P')
    })

    it('applies correct CSS classes to subtitle', () => {
      render(<HeaderLogo />)
      
      const subtitle = screen.getByText('What the fork is for dinner?')
      expect(subtitle).toHaveClass(
        'text-lg',
        'text-muted',
        'font-light',
        'italic'
      )
    })

    it('has correct container structure', () => {
      const { container } = render(<HeaderLogo />)
      
      const containerDiv = container.querySelector('.container')
      expect(containerDiv).toBeInTheDocument()
      expect(containerDiv).toHaveClass(
        'container',
        'mx-auto',
        'px-4',
        'py-6'
      )
      
      const titleSection = container.querySelector('.text-center')
      expect(titleSection).toBeInTheDocument()
    })
  })

  describe('Navigation Section', () => {
    it('renders navigation element', () => {
      render(<HeaderLogo />)
      
      const nav = screen.getByRole('navigation')
      expect(nav).toBeInTheDocument()
    })

    it('applies correct CSS classes to navigation', () => {
      render(<HeaderLogo />)
      
      const nav = screen.getByRole('navigation')
      expect(nav).toHaveClass('flex', 'justify-center', 'mt-4')
    })

    it('has correct navigation container structure', () => {
      const { container } = render(<HeaderLogo />)
      
      const navContainer = container.querySelector('nav .flex.space-x-8')
      expect(navContainer).toBeInTheDocument()
      expect(navContainer).toHaveClass('flex', 'space-x-8')
    })
  })

  describe('Navigation Links', () => {
    it('renders Home link with correct attributes', () => {
      render(<HeaderLogo />)
      
      const homeLink = screen.getByTestId('link-/')
      expect(homeLink).toBeInTheDocument()
      expect(homeLink).toHaveAttribute('href', '/')
      expect(homeLink).toHaveTextContent('Home')
    })

    it('renders Shopping list link with correct attributes', () => {
      render(<HeaderLogo />)
      
      const shoppingLink = screen.getByTestId('link-/shopping')
      expect(shoppingLink).toBeInTheDocument()
      expect(shoppingLink).toHaveAttribute('href', '/shopping')
      expect(shoppingLink).toHaveTextContent('Shopping list')
    })

    it('renders Past plans link with correct attributes', () => {
      render(<HeaderLogo />)
      
      const pastPlansLink = screen.getByTestId('link-/weeks')
      expect(pastPlansLink).toBeInTheDocument()
      expect(pastPlansLink).toHaveAttribute('href', '/weeks')
      expect(pastPlansLink).toHaveTextContent('Past plans')
    })

    it('applies correct CSS classes to all navigation links', () => {
      render(<HeaderLogo />)
      
      const expectedClasses = [
        'text-secondary',
        'hover:text-foreground',
        'transition-colors',
        'font-medium',
        'underline-offset-4',
        'hover:underline'
      ]
      
      const homeLink = screen.getByTestId('link-/')
      const shoppingLink = screen.getByTestId('link-/shopping')
      const pastPlansLink = screen.getByTestId('link-/weeks')
      
      expectedClasses.forEach(className => {
        expect(homeLink).toHaveClass(className)
        expect(shoppingLink).toHaveClass(className)
        expect(pastPlansLink).toHaveClass(className)
      })
    })

    it('renders exactly three navigation links', () => {
      render(<HeaderLogo />)
      
      const allLinks = screen.getAllByRole('link')
      expect(allLinks).toHaveLength(3)
    })

    it('renders links in correct order', () => {
      render(<HeaderLogo />)
      
      const allLinks = screen.getAllByRole('link')
      expect(allLinks[0]).toHaveTextContent('Home')
      expect(allLinks[1]).toHaveTextContent('Shopping list')
      expect(allLinks[2]).toHaveTextContent('Past plans')
    })
  })

  describe('Accessibility', () => {
    it('has semantic HTML structure', () => {
      render(<HeaderLogo />)
      
      // Check for semantic elements
      expect(screen.getByRole('banner')).toBeInTheDocument() // header
      expect(screen.getByRole('navigation')).toBeInTheDocument() // nav
      expect(screen.getByRole('heading', { level: 1 })).toBeInTheDocument() // h1
    })

    it('has accessible link text', () => {
      render(<HeaderLogo />)
      
      const homeLink = screen.getByRole('link', { name: 'Home' })
      const shoppingLink = screen.getByRole('link', { name: 'Shopping list' })
      const pastPlansLink = screen.getByRole('link', { name: 'Past plans' })
      
      expect(homeLink).toBeInTheDocument()
      expect(shoppingLink).toBeInTheDocument()
      expect(pastPlansLink).toBeInTheDocument()
    })
  })

  describe('Layout and Styling', () => {
    it('applies responsive text sizing classes', () => {
      render(<HeaderLogo />)
      
      const title = screen.getByRole('heading', { level: 1 })
      expect(title).toHaveClass('text-4xl', 'md:text-5xl')
    })

    it('has correct spacing and layout classes', () => {
      const { container } = render(<HeaderLogo />)
      
      // Check container spacing
      const containerDiv = container.querySelector('.container')
      expect(containerDiv).toHaveClass('px-4', 'py-6')
      
      // Check title section centering
      const titleSection = container.querySelector('.text-center')
      expect(titleSection).toBeInTheDocument()
      
      // Check navigation centering and spacing
      const nav = screen.getByRole('navigation')
      expect(nav).toHaveClass('flex', 'justify-center', 'mt-4')
    })

    it('applies hover and transition classes to links', () => {
      render(<HeaderLogo />)
      
      const links = screen.getAllByRole('link')
      
      links.forEach(link => {
        expect(link).toHaveClass(
          'hover:text-foreground',
          'transition-colors',
          'hover:underline'
        )
      })
    })
  })

  describe('Content Verification', () => {
    it('contains expected text content', () => {
      render(<HeaderLogo />)
      
      expect(screen.getByText('Family Foodie')).toBeInTheDocument()
      expect(screen.getByText('What the fork is for dinner?')).toBeInTheDocument()
      expect(screen.getByText('Home')).toBeInTheDocument()
      expect(screen.getByText('Shopping list')).toBeInTheDocument()
      expect(screen.getByText('Past plans')).toBeInTheDocument()
    })

    it('matches snapshot', () => {
      const { container } = render(<HeaderLogo />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('Integration Tests', () => {
    it('works properly when rendered multiple times', () => {
      const { unmount } = render(<HeaderLogo />)
      expect(screen.getByText('Family Foodie')).toBeInTheDocument()
      
      unmount()
      
      render(<HeaderLogo />)
      expect(screen.getByText('Family Foodie')).toBeInTheDocument()
    })

    it('maintains component structure integrity', () => {
      render(<HeaderLogo />)
      
      // Verify complete component hierarchy exists
      const header = screen.getByRole('banner')
      const container = header.querySelector('.container')
      const titleSection = container.querySelector('.text-center')
      const nav = screen.getByRole('navigation')
      const navContainer = nav.querySelector('.flex.space-x-8')
      
      expect(header).toContainElement(container)
      expect(container).toContainElement(titleSection)
      expect(container).toContainElement(nav)
      expect(nav).toContainElement(navContainer)
    })
  })
})