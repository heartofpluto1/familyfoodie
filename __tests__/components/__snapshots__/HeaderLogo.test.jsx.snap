// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HeaderLogo Component Content Verification matches snapshot 1`] = `
<header
  class="bg-surface border-b border-custom"
>
  <div
    class="container mx-auto px-4 py-6"
  >
    <div
      class="text-center"
    >
      <h1
        class="mocked-crimson-text-font font-bold text-4xl md:text-5xl text-foreground mb-2 tracking-wide"
      >
        Family Foodie
      </h1>
      <p
        class="text-lg text-muted font-light italic"
      >
        What the fork is for dinner?
      </p>
    </div>
    <nav
      class="flex justify-center mt-4"
    >
      <div
        class="flex space-x-8"
      >
        <a
          class="text-secondary hover:text-foreground transition-colors font-medium underline-offset-4 hover:underline"
          data-testid="link-/"
          href="/"
        >
          Home
        </a>
        <a
          class="text-secondary hover:text-foreground transition-colors font-medium underline-offset-4 hover:underline"
          data-testid="link-/shopping"
          href="/shopping"
        >
          Shopping list
        </a>
        <a
          class="text-secondary hover:text-foreground transition-colors font-medium underline-offset-4 hover:underline"
          data-testid="link-/weeks"
          href="/weeks"
        >
          Past plans
        </a>
      </div>
    </nav>
  </div>
</header>
`;
